core/src/test/java/io/grpc/AbstractForwardingTest.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2017, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import static junit.framework.TestCase.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mockingDetails;
+import static org.mockito.Mockito.verify;
+
+import com.google.common.base.Defaults;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * An abstract base class for testing forwarding classes. This automatically checks that
+ * all public methods of the {@link #delegateClass()} are forwarded by the forwarder to the
+ * delegate. This does NOT verify that arguments are forwarded properly. It only alerts
+ * the developer if a forward method is missing.
+ */
+@RunWith(JUnit4.class)
+public abstract class AbstractForwardingTest<T> {
+  /**
+   * Returns a mock object that can be used with {@code Mockito.verify()}.
+   */
+  public abstract T mockDelegate();
```

Comment:
`protected`? because its only for overriding, not for invocation outside this class.
Ditto for below.